tests/testthat/helper-expect-metadata.R

```diff
@@ -0,0 +1,246 @@
+      options = list(
+        list(
+          id = scalar("opt1"),
+          label = scalar("Option 1")
+        ),
+        list(
+          id = scalar("opt2"),
+          label = scalar("Option 2")
+        )
+      )
+    )
+  })
+}
+
+## Test validity of the metadata, namely that ids used in any effect which
+## reference filters actually exist in the filter types
+expect_valid_metadata <- function(metadata) {
+  filter_ids <- lapply(metadata$filterTypes, "[[", "id")
+  filter_types <- setNames(metadata$filterTypes, filter_ids)
+
+  plot_names <- names(metadata$plotSettingsControl)
+  for (plot in plot_names) {
+    default_effects <- metadata$plotSettingsControl[[plot]]$defaultEffect
+    default_filter_map <- validate_effects(
+      filter_types,
+      metadata$plotSettingsControl[[plot]]$defaultEffect,
+      paste0("plot: '", plot, "', default effects"))
+    for (plot_setting in metadata$plotSettingsControl[[plot]]$plotSettings) {
+      for (option in plot_setting$options) {
+        validate_effects(
+          filter_types,
+          plot_setting$options$effect,
+          paste0("plot: '",  plot,
+                 "', setting: '", plot_setting$id,
+                 "', option: '", option$id, "'"),
+          default_filter_map)
+      }
+    }
+  }
+  invisible(TRUE)
+}
+
+#' Validate effects
+#'
+#' Effects have filterId and stateFilterId
+#' stateFilterId is the unique ID for the dropdown
+#' filterId is the unique ID of the FilterType, used to identify options
+#' this is because there can be multiple dropdowns using the same underlying
+#' data e.g. colour and size indicator in the bubble plot
+#' setFilter effects have the mapping between filterId and stateFilterId
+#' all other effects refer only to the stateFilterId
+#'
+#' @param filter_types Named filter types
+#' @param effects Effects to check
+#' @param context Additional context appended to any test failure messages
+#' @param state_filter_map Map of state filter ID to filter ID. See above for
+#'   details. Used to check that effects refer to a filter which exists.
+#'
+#' @return Nothing, called for side effect
+#' @noRd
+validate_effects <- function(filter_types, effects, context,
+                             state_filter_map = list()) {
+  if (!is.null(effects$setFilters)) {
+    filter_ids <- vcapply(effects$setFilters, "[[", "filterId")
+    expect_filter_exists(names(filter_types),
+                         filter_ids,
+                         paste(context, "setFilters"))
+    state_filter_ids <- vcapply(effects$setFilters, "[[", "stateFilterId")
+    state_filter_map <- c(state_filter_map,
+                          setNames(filter_ids, state_filter_ids))
+    expect_state_filter_id_unique(state_filter_map)
+  } else if (length(state_filter_map) == 0) {
+    ## This is kind of gross but logic is, if setFilters is not included
+    ## in an effect, then use all the available filters. So have to replicate
+    ## this here for validation
```

Comment:
@M-Kusumgar this is true right?

Reply:
Thanks! Should have looked at that code, thanks for the link. I've updated the test using proper information from here. This has simplified this function a bit too!